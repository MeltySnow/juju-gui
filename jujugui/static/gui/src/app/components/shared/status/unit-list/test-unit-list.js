/* Copyright (C) 2018 Canonical Ltd. */
'use strict';

const React = require('react');
const enzyme = require('enzyme');

const StatusUnitList = require('./unit-list');

describe('StatusUnitList', () => {
  let applications, units;

  const renderComponent = (options = {}) => enzyme.shallow(
    <StatusUnitList
      applications={options.applications || applications}
      changeState={options.changeState || sinon.stub()}
      generateMachineURL={options.generateCharmURL || sinon.stub()}
      generatePath={options.generatePath || sinon.stub()}
      generateUnitOnClick={
        options.generateUnitOnClick || sinon.stub().returns(sinon.stub())}
      generateUnitURL={
        options.generateUnitURL || sinon.stub().returns('http://example.com')}
      onMachineClick={options.onCharmClick || sinon.stub()}
      statusFilter={options.statusFilter}
      units={options.units || units} />
  );

  beforeEach(() => {
    applications = {
      getById: sinon.stub().returns({
        get: sinon.stub().withArgs('exposed').returns(false)
      })
    };
    units = [{
      agentStatus: 'idle',
      displayName: 'django/0',
      id: 'django/id0',
      machine: '1',
      public_address: '1.2.3.4',
      portRanges: [
        {from: 80, to: 80, protocol: 'tcp'},
        {from: 443, to: 443, protocol: 'tcp'}
      ],
      workloadStatus: 'installing',
      workloadStatusMessage: 'these are the voyages'
    }, {
      agentStatus: 'executing',
      displayName: 'django/1',
      id: 'django/id1',
      machine: '2',
      public_address: '1.2.3.5',
      portRanges: [
        {from: 80, to: 88, protocol: 'udp'}
      ],
      workloadStatus: 'error',
      workloadStatusMessage: 'exterminate!'
    }, {
      agentStatus: 'idle',
      displayName: 'django/42',
      id: 'django/id42',
      machine: '2',
      public_address: '1.2.3.6',
      // Simulate that the unit didn't open ports yet.
      portRanges: [],
      workloadStatus: 'installing',
      workloadStatusMessage: ''
    }, {
      // Unplaced units are excluded.
      agentStatus: '',
      displayName: 'django/2',
      id: 'django/id2',
      public_address: '',
      portRanges: [],
      workloadStatus: '',
      workloadStatusMessage: ''
    }, {
      // Uncommitted units are excluded.
      agentStatus: '',
      displayName: 'django/3',
      id: 'django/id3',
      machine: 'new42',
      public_address: '',
      portRanges: [],
      workloadStatus: '',
      workloadStatusMessage: ''
    }];
  });

  it('renders', () => {
    const wrapper = renderComponent();
    expect(wrapper).toMatchSnapshot();
  });

  it('displays a link for exposed units', () => {
    applications.getById.returns({
      get: sinon.stub().withArgs('exposed').returns(true)
    });
    const wrapper = renderComponent();
    assert.equal(
      wrapper.prop('rows')[0].columns[4].content.props.className.includes('status-view__link'),
      true);
  });
});
