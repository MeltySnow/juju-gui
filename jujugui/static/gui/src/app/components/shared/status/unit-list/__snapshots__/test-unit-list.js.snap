// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusUnitList renders 1`] = `
<StatusTable
  changeState={[Function]}
  generatePath={[Function]}
  headers={
    Array [
      Object {
        "columnSize": 2,
        "content": "Unit",
      },
      Object {
        "columnSize": 2,
        "content": "Workload",
      },
      Object {
        "columnSize": 2,
        "content": "Agent",
      },
      Object {
        "columnSize": 1,
        "content": "Machine",
      },
      Object {
        "columnSize": 2,
        "content": "Public address",
      },
      Object {
        "columnSize": 1,
        "content": "Ports",
      },
      Object {
        "columnSize": 2,
        "content": "Message",
      },
    ]
  }
  rows={
    Array [
      Object {
        "classes": Array [
          "status-table__row--pending",
        ],
        "clickState": Object {
          "app": "state",
        },
        "columns": Array [
          Object {
            "columnSize": 2,
            "content": <span>
              <img
                className="status-view__icon"
                src={false}
              />
              django/0
            </span>,
          },
          Object {
            "columnSize": 2,
            "content": <StatusLabel
              status="installing"
            />,
          },
          Object {
            "columnSize": 2,
            "content": <StatusLabel
              status="idle"
            />,
          },
          Object {
            "columnSize": 1,
            "content": <a
              className="status-view__link"
              onClick={[Function]}
            >
              1
            </a>,
          },
          Object {
            "columnSize": 2,
            "content": "1.2.3.4",
          },
          Object {
            "columnSize": 1,
            "content": "80/tcp, 443/tcp",
          },
          Object {
            "columnSize": 2,
            "content": "these are the voyages",
          },
        ],
        "extraData": "pending",
        "key": "django/id0",
      },
      Object {
        "classes": Array [
          "status-table__row--error",
        ],
        "clickState": Object {
          "app": "state",
        },
        "columns": Array [
          Object {
            "columnSize": 2,
            "content": <span>
              <img
                className="status-view__icon"
                src={false}
              />
              django/1
            </span>,
          },
          Object {
            "columnSize": 2,
            "content": <StatusLabel
              status="error"
            />,
          },
          Object {
            "columnSize": 2,
            "content": <StatusLabel
              status="executing"
            />,
          },
          Object {
            "columnSize": 1,
            "content": <a
              className="status-view__link"
              onClick={[Function]}
            >
              2
            </a>,
          },
          Object {
            "columnSize": 2,
            "content": "1.2.3.5",
          },
          Object {
            "columnSize": 1,
            "content": "80-88/udp",
          },
          Object {
            "columnSize": 2,
            "content": "exterminate!",
          },
        ],
        "extraData": "error",
        "key": "django/id1",
      },
      Object {
        "classes": Array [
          "status-table__row--pending",
        ],
        "clickState": Object {
          "app": "state",
        },
        "columns": Array [
          Object {
            "columnSize": 2,
            "content": <span>
              <img
                className="status-view__icon"
                src={false}
              />
              django/42
            </span>,
          },
          Object {
            "columnSize": 2,
            "content": <StatusLabel
              status="installing"
            />,
          },
          Object {
            "columnSize": 2,
            "content": <StatusLabel
              status="idle"
            />,
          },
          Object {
            "columnSize": 1,
            "content": <a
              className="status-view__link"
              onClick={[Function]}
            >
              2
            </a>,
          },
          Object {
            "columnSize": 2,
            "content": "1.2.3.6",
          },
          Object {
            "columnSize": 1,
            "content": "",
          },
          Object {
            "columnSize": 2,
            "content": "",
          },
        ],
        "extraData": "pending",
        "key": "django/id42",
      },
    ]
  }
/>
`;
